File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableSchema.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.table.api

import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.api.common.typeutils.CompositeType

import _root_.scala.collection.mutable.ArrayBuffer
import _root_.java.util.Objects

import org.apache.flink.table.calcite.FlinkTypeFactory

/**
  * A TableSchema represents a Table's structure.
  */
class TableSchema(
  private val columnNames: Array[String],
  private val columnTypes: Array[TypeInformation[_]]) {

  private val columnNameToIndex: Map[String, Int] = columnNames.zipWithIndex.toMap

  if (columnNames.length != columnTypes.length) {
    throw new TableException(
      s"Number of field names and field types must be equal.\n" +
        s"Number of names is ${columnNames.length}, number of types is ${columnTypes.length}.\n" +
        s"List of field names: ${columnNames.mkString("[", ", ", "]")}.\n" +
        s"List of field types: ${columnTypes.mkString("[", ", ", "]")}.")
  }

  // check uniqueness of field names
  if (columnNames.toSet.size != columnTypes.length) {
    val duplicateFields = columnNames
      // count occurrences of field names
      .groupBy(identity).mapValues(_.length)
      // filter for occurrences > 1 and map to field name
      .filter(g => g._2 > 1).keys

    throw new TableException(
      s"Field names must be unique.\n" +
        s"List of duplicate fields: ${duplicateFields.mkString("[", ", ", "]")}.\n" +
        s"List of all fields: ${columnNames.mkString("[", ", ", "]")}.")
  }

  /**
    * Returns a deep copy of the TableSchema.
    */
  def copy: TableSchema = {
    new TableSchema(columnNames.clone(), columnTypes.clone())
  }

  /**
    * Returns all type information as an array.
    */
  def getTypes: Array[TypeInformation[_]] = columnTypes

  /**
    * Returns the specified type information for the given column index.
    *
    * @param columnIndex the index of the field
    */
  def getType(columnIndex: Int): Option[TypeInformation[_]] = {
    if (columnIndex < 0 || columnIndex >= columnNames.length) {
      None
    } else {
      Some(columnTypes(columnIndex))
    }
  }

  /**
    * Returns the specified type information for the given column name.
    *
    * @param columnName the name of the field
    */
  def getType(columnName: String): Option[TypeInformation[_]] = {
    if (columnNameToIndex.contains(columnName)) {
      Some(columnTypes(columnNameToIndex(columnName)))
    } else {
      None
    }
  }

  /**
    * Returns the number of columns.
    */
  def getColumnCount: Int = columnNames.length

  /**
    * Returns all column names as an array.
    */
  def getColumnNames: Array[String] = columnNames

  /**
    * Returns the specified column name for the given column index.
    *
    * @param columnIndex the index of the field
    */
  def getColumnName(columnIndex: Int): Option[String] = {
    if (columnIndex < 0 || columnIndex >= columnNames.length) {
      None
    } else {
      Some(columnNames(columnIndex))
    }
  }

  /**
    * Converts a table schema into a schema that represents the result that would be written
    * into a table sink or operator outside of the Table & SQL API. Time attributes are replaced
    * by proper TIMESTAMP data types.
    *
    * @return a table schema with no time attributes
    */
  def withoutTimeAttributes: TableSchema = {
    val converted = columnTypes.map { t =>
      if (FlinkTypeFactory.isTimeIndicatorType(t)) {
        Types.SQL_TIMESTAMP
      } else {
        t
      }
    }
    new TableSchema(columnNames, converted)
  }

  override def toString: String = {
    val builder = new StringBuilder
    builder.append("root\n")
    columnNames.zip(columnTypes).foreach{ case (name, typeInfo) =>
      builder.append(s" |-- $name: $typeInfo\n")
    }
    builder.toString()
  }

  override def equals(other: Any): Boolean = {
    other match {
      case that: TableSchema if that.canEqual(this) =>
        this.columnNames.deep == that.columnNames.deep &&
        this.columnTypes.deep == that.columnTypes.deep
      case _ => false
    }
  }

  def canEqual(other: Any): Boolean = other.isInstanceOf[TableSchema]

  override def hashCode(): Int = {
    Objects.hash(columnNames, columnTypes)
  }
}

object TableSchema {

  /**
    * Creates a [[TableSchema]] from a [[TypeInformation]].
    * If the [[TypeInformation]] is a [[CompositeType]], the fieldnames and types for the composite
    * type are used to construct the [[TableSchema]].
    * Otherwise, a [[TableSchema]] with a single field is created. The field name is "f0" and the
    * field type the provided type.
    *
    * @param typeInfo The [[TypeInformation]] from which the [[TableSchema]] is generated.
    * @return The [[TableSchema]] that was generated from the given [[TypeInformation]].
    */
  def fromTypeInfo(typeInfo: TypeInformation[_]): TableSchema = {
    typeInfo match {
      case c: CompositeType[_] =>
        // get field names and types from composite type
        val fieldNames = c.getFieldNames
        val fieldTypes = fieldNames.map(c.getTypeAt).asInstanceOf[Array[TypeInformation[_]]]
        new TableSchema(fieldNames, fieldTypes)
      case t: TypeInformation[_] =>
        // create table schema with a single field named "f0" of the given type.
        new TableSchema(Array("f0"), Array(t))
    }
  }

  def builder(): TableSchemaBuilder = {
    new TableSchemaBuilder
  }

}

class TableSchemaBuilder {

  private val fieldNames: ArrayBuffer[String] = new ArrayBuffer[String]()
  private val fieldTypes: ArrayBuffer[TypeInformation[_]] = new ArrayBuffer[TypeInformation[_]]()

  def field(name: String, tpe: TypeInformation[_]): TableSchemaBuilder = {
    fieldNames.append(name)
    fieldTypes.append(tpe)
    this
  }

  def build(): TableSchema = {
    new TableSchema(fieldNames.toArray, fieldTypes.toArray)
  }
}

File: flink-python/pyflink/table/table_schema.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
import sys

from pyflink.java_gateway import get_gateway
from pyflink.table.types import _to_java_type, _from_java_type
from pyflink.util.utils import to_jarray

if sys.version >= '3':
    unicode = str

__all__ = ['TableSchema']


class TableSchema(object):
    """
    A table schema that represents a table's structure with field names and data types.
    """

    def __init__(self, field_names=None, data_types=None, j_table_schema=None):
        if j_table_schema is None:
            gateway = get_gateway()
            j_field_names = to_jarray(gateway.jvm.String, field_names)
            j_data_types = to_jarray(gateway.jvm.TypeInformation,
                                     [_to_java_type(item) for item in data_types])
            self._j_table_schema = gateway.jvm.TableSchema(j_field_names, j_data_types)
        else:
            self._j_table_schema = j_table_schema

    def copy(self):
        """
        Returns a deep copy of the table schema.

        :return: A deep copy of the table schema.
        """
        return TableSchema(j_table_schema=self._j_table_schema.copy())

    def get_field_data_types(self):
        """
        Returns all field data types as a list.

        :return: A list of all field data types.
        """
        return [_from_java_type(item) for item in self._j_table_schema.getFieldDataTypes()]

    def get_field_data_type(self, field):
        """
        Returns the specified data type for the given field index or field name.

        :param field: The index of the field or the name of the field.
        :return: The data type of the specified field.
        """
        if not isinstance(field, (int, str, unicode)):
            raise TypeError("Expected field index or field name, got %s" % type(field))
        optional_result = self._j_table_schema.getFieldDataType(field)
        if optional_result.isPresent():
            return _from_java_type(optional_result.get())
        else:
            return None

    def get_field_count(self):
        """
        Returns the number of fields.

        :return: The number of fields.
        """
        return self._j_table_schema.getFieldCount()

    def get_field_names(self):
        """
        Returns all field names as a list.

        :return: The list of all field names.
        """
        return list(self._j_table_schema.getFieldNames())

    def get_field_name(self, field_index):
        """
        Returns the specified name for the given field index.

        :param field_index: The index of the field.
        :return: The field name.
        """
        optional_result = self._j_table_schema.getFieldName(field_index)
        if optional_result.isPresent():
            return optional_result.get()
        else:
            return None

    def to_row_data_type(self):
        """
        Converts a table schema into a (nested) data type describing a
        :func:`pyflink.table.types.DataTypes.ROW`.

        :return: The row data type.
        """
        return _from_java_type(self._j_table_schema.toRowDataType())

    def __repr__(self):
        return self._j_table_schema.toString()

    def __eq__(self, other):
        return isinstance(other, self.__class__) and self._j_table_schema == other._j_table_schema

    def __hash__(self):
        return self._j_table_schema.hashCode()

    def __ne__(self, other):
        return not self.__eq__(other)

    @classmethod
    def builder(cls):
        return TableSchema.Builder()

    class Builder(object):
        """
        Builder for creating a :class:`TableSchema`.
        """

        def __init__(self):
            self._field_names = []
            self._field_data_types = []

        def field(self, name, data_type):
            """
            Add a field with name and data type.

            The call order of this method determines the order of fields in the schema.

            :param name: The field name.
            :param data_type: The field data type.
            :return: This object.
            """
            assert name is not None
            assert data_type is not None
            self._field_names.append(name)
            self._field_data_types.append(data_type)
            return self

        def build(self):
            """
            Returns a :class:`TableSchema` instance.

            :return: The :class:`TableSchema` instance.
            """
            return TableSchema(self._field_names, self._field_data_types)
